node/pkg/accountant/submit_obs.go

```diff
@@ -205,69 +304,73 @@ func SubmitObservationToContract(
-func CheckSubmitObservationResult(txResp *sdktx.BroadcastTxResponse) (bool, error) {
 	if txResp == nil {
-		return false, fmt.Errorf("txResp is nil")
+		return txResp, fmt.Errorf("sent broadcast but returned txResp is nil")
 	}
+
 	if txResp.TxResponse == nil {
-		return false, fmt.Errorf("txResp does not contain a TxResponse")
+		return txResp, fmt.Errorf("sent broadcast but returned txResp.TxResponse is nil")
 	}
+
 	if txResp.TxResponse.RawLog == "" {
-		return false, fmt.Errorf("RawLog is not set")
+		return txResp, fmt.Errorf("sent broadcast but raw_log is not set, unable to analyze the result")
 	}
-	if strings.Contains(txResp.TxResponse.RawLog, "execute wasm contract failed") {
-		if strings.Contains(txResp.TxResponse.RawLog, "already committed") {
-			return true, nil
 
-		}
+	if strings.Contains(txResp.TxResponse.RawLog, "out of gas") {
+		return txResp, fmt.Errorf("out of gas: %s", txResp.TxResponse.RawLog)
+	}
 
-		// TODO Need to test this, requires multiple guardians.
-		if strings.Contains(txResp.TxResponse.RawLog, "duplicate signature") {
-			return false, nil
-		}
+	if strings.Contains(txResp.TxResponse.RawLog, "failed to execute message") {
+		return txResp, fmt.Errorf("failed to submit observations: %s", txResp.TxResponse.RawLog)
+	}
+
+	logger.Debug("acct: in SubmitObservationsToContract, done sending broadcast", zap.String("resp", wormchainConn.BroadcastTxResponseToString(txResp)))
+	return txResp, nil
+}
+
+// TODO: Once the issue with the junk characters is sorted out, drop the logger parameter.
 
-		return false, fmt.Errorf(txResp.TxResponse.RawLog)
+// GetObservationResponses is a free function that extracts the observation responses from a transaction response.
+// It assumes the transaction response is valid (SubmitObservationsToContract() did not return an error).
+func GetObservationResponses(logger *zap.Logger, txResp *sdktx.BroadcastTxResponse, numExpected int) (ObservationResponses, error) {
+	data, err := hex.DecodeString(txResp.TxResponse.Data)
+	if err != nil {
+		return nil, fmt.Errorf("failed to decode data: %w", err)
 	}
 
-	if strings.Contains(txResp.TxResponse.RawLog, "failed to execute message") {
-		return false, fmt.Errorf(txResp.TxResponse.RawLog)
+	var msg sdktypes.TxMsgData
+	if err := msg.Unmarshal([]byte(data)); err != nil {
+		return nil, fmt.Errorf("failed to unmarshal data: %w", err)
+	}
+
+	str := string(msg.Data[0].Data)
+
+	idx := strings.Index(str, "[")
+	if idx >= 0 {
+		logger.Info("Why does the data start with this?", zap.String("junk", str[:idx]))
+		str = str[idx:]
+	}
+
+	var responses ObservationResponses
+	err = json.Unmarshal([]byte(str), &responses)
+	if err != nil {
+		return nil, fmt.Errorf("failed to unmarshal responses: %w", err)
+	}
+
+	if len(responses) != numExpected {
+		return nil, fmt.Errorf("unexpected number of responses, expected %d, actual %d", numExpected, len(responses))
 	}
 
-	return false, nil
+	return responses, nil
```

Comment:
If we wanted to return a map rather than a slice we just need one more step here:

```go
out := make(map[string]ObservationResponseStatus)
for _, r := responses {
	out[r.Key.String()] = r.Status
}

return out, nil
```

Reply:
Changed.